gadget/install/install.go

```diff
@@ -306,21 +307,86 @@ func FactoryReset(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string,
 		return nil, fmt.Errorf("cannot read %v partitions: %v", bootDevice, err)
 	}
 
-	// verify that layout is compatible, note this will fail if partition
-	// table is corrupted and the expected partitions are not present
+	roleNeedsEncryption := func(role string) bool {
+		return role == gadget.SystemData || role == gadget.SystemSave
+	}
+
 	layoutCompatOps := &gadget.EnsureLayoutCompatibilityOptions{
 		AssumeCreatablePartitionsCreated: true,
+		ExpectedStructureEncryption:      map[string]gadget.StructureEncryptionParameters{},
+	}
+	if options.EncryptionType != secboot.EncryptionTypeNone {
+		var encryptionParam gadget.StructureEncryptionParameters
+		switch options.EncryptionType {
+		case secboot.EncryptionTypeLUKS:
+			encryptionParam = gadget.StructureEncryptionParameters{Method: gadget.EncryptionLUKS}
+		default:
+			// XXX what about ICE?
+			return nil, fmt.Errorf("unsupported encryption type %v", options.EncryptionType)
+		}
+		for _, volStruct := range laidOutBootVol.LaidOutStructure {
+			if !roleNeedsEncryption(volStruct.Role) {
+				continue
+			}
+			if layoutCompatOps.ExpectedStructureEncryption == nil {
+				layoutCompatOps.ExpectedStructureEncryption = map[string]gadget.StructureEncryptionParameters{}
+			}
+			layoutCompatOps.ExpectedStructureEncryption[volStruct.Name] = encryptionParam
+		}
 	}
+	// factory reset is done on a system that was once installed, so this
+	// should be always successful unless the partition table has changed
 	if err := gadget.EnsureLayoutCompatibility(laidOutBootVol, diskLayout, layoutCompatOps); err != nil {
 		return nil, fmt.Errorf("gadget and system-boot device %v partition table not compatible: %v", bootDevice, err)
 	}
 
-	hasSavePartition := len(partitionsWithRolesAndContent(laidOutBootVol, diskLayout, []string{gadget.SystemSave})) != 0
+	var keyForRole map[string]keys.EncryptionKey
+	var deviceForRole = map[string]string{}
```

Comment:
Just double-check, since I recall that in some standup it was mentioned the case of some customer having more than a data partition: should this be a `map[string][]string` instead? to support more than one partition for the same role?

Reply:
I've added some checks to make sure that only devices corresponding to structures with roles are in the map. This should address it since it's not possible to have more than one structure with a given role.